dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -230,18 +235,31 @@ class NVDECCache {
     }
 
  private:
-    NVDECCache() {}
+    NVDECCache() {
+      int num_devices = 0;
+      CUDA_CALL(cudaGetDeviceCount(&num_devices));
+      dec_cache.resize(num_devices);
+      access_lock.resize(num_devices);
+    }
 
     ~NVDECCache() {
-      std::scoped_lock lock(access_lock);
-      for (auto &it : dec_cache) {
-        cuvidDestroyDecoder(it.second.decoder);
+      for (size_t i = 0; i < dec_cache.size(); ++i) {
+        auto &dec_cache_elm = dec_cache[i];
+        std::scoped_lock lock(access_lock[i]);
```

Comment:
What do you need a `scoped_lock` for? I don't think it has any advantage over `lock_guard` when locking on e object.

Reply:
Fixed